src/actions/mod.rs

```diff
@@ -197,21 +197,81 @@ impl InitActionContext {
     }
 
     fn build<O: Output>(&self, project_path: &Path, priority: BuildPriority, out: O) {
-        struct BuildNotifier<O: Output> {
+
+        lazy_static! {
+            static ref PROGRESS_ID_COUNTER: AtomicUsize = {
+                AtomicUsize::new(0)
+            };
+        }
+
+        fn new_progress_params(id: String, title: &str) -> ProgressParams {
+            ProgressParams {
+                id,
+                title: Some(title.into()),
+                message: None,
+                percentage: None,
+                done: None,
+            }
+        }
+
+        // notifier of diagnostics after the build
+        struct BuildDiagnosticsNotifier<O: Output> {
             out: O,
             active_build_count: Arc<AtomicUsize>,
+            progress_params: ProgressParams,
         }
 
-        impl<O: Output> Notifier for BuildNotifier<O> {
-            fn notify_begin(&self) {
-                self.out.notify(Notification::<DiagnosticsBegin>::new(NoParams {}));
+        // base progress parameters for notification of the analysis.
+        let diganostics_progress_params = {
+            let id = format!("diagnostics_{}", PROGRESS_ID_COUNTER.fetch_add(1, Ordering::SeqCst));
```

Comment:
re generating ID inside `fn new_progress_params`, I'd stick to just generating a number, since it's more of a implementation detail and we already distinguish using progress `title`s